docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,174 @@
+// by the nodeSelectorTerms.
+// TODO: what are the semantics of empty and null?
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
+	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms,omitempty"`
+}
+
+// An empty node selector term matches all objects. A null node selector term
+// matches no objects.
+type NodeSelectorTerm struct {
+	// matchExpressions is a list of node selector requirements. The requirements are ANDed.
+	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// A node selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values.
+type NodeSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators ard In, NotIn, Exists, DoesNotExist. Gt, and Lt.
+	Operator NodeSelectorOperator `json:"operator"`
+	// values is an array of string values. If the operator is In or NotIn,
+	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
+	// the values array must be empty. If the operator is Gt or Lt, the values
+	// array must have a single element, which will be interpreted as an integer.
+    // This array is replaced during a strategic merge patch.
+	Values []string `json:"values,omitempty"`
+}
+
+// A node selector operator is the set of operators that can be used in
+// a node selector requirement.
+type NodeSelectorOperator string
+
+const (
+	NodeSelectorOpIn           NodeSelectorOperator = "In"
+	NodeSelectorOpNotIn        NodeSelectorOperator = "NotIn"
+	NodeSelectorOpExists       NodeSelectorOperator = "Exists"
+	NodeSelectorOpDoesNotExist NodeSelectorOperator = "DoesNotExist"
+	NodeSelectorOpGt           NodeSelectorOperator = "Gt"
+	NodeSelectorOpLt           NodeSelectorOperator = "Lt"
+)
+```
+
+### Changes to PodSpec
+
+We will add two fields to `PodSpec`:
+
+```go
+// NodeConstraints specifies hard scheduling requirements for the Pod.
+NodeConstraints *NodeSelector `json:nodeConstraints,omitempty"`
+// NodePreferences specifies which nodes are preferred among those that meet all
+// of the hard scheduling requirements (resource request, constraints, etc.).
+// We don't use *NodeSelector as the type here, so as to make it clear that
+// we are not producing a single set of nodes as the result, but rather describing
+// several sets of nodes. A node that is in more such sets is
+// considered to be "more preferred" than a node that is in fewer.
+NodePreferences []NodeSelectorTerm `json:nodePreferences,omitempty"`
+```
+
+Unfortunately, the name of the existing `map[string]string` field in PodSpec is `NodeSelector`
+and we can't change it since this name is part of the API. Hopefully this won't
+cause too much confusion.
```

Comment:
I think it will cause confusion having the same term `NodeSelector` used in two different places in Kubernetes meaning two different things. We should work very hard to avoid this.  
The `NodeAffinity` proposal above is one way of solving this.